### PR TITLE
replace logging's debug function

### PR DESCRIPTION
**Pull request by dwpeng**
#567

### CODE REVIEW

**Comment by dwpeng** [comment]
#568

File: jcvi/apps/base.py [hunk 1]
```diff
@@ -37,6 +37,20 @@
 os.environ["LC_ALL"] = "C"
 JCVIHELP = "JCVI utility libraries {} [{}]\n".format(__version__, __copyright__)
 
+_logging_debug = logging.debug
+def debug_with_location(message: str):
```
Comment (tanghaibao) [inline jcvi/apps/base.py:41]:
```suggestion
def debug_with_location(message: str, *args: Any):
```

Comment (tanghaibao) [reply jcvi/apps/base.py:41]: Test is failing due to many places invoking with multiple args, e.g. `logging.debug("message: %s", message)`

**Comment by dwpeng** [comment]
Thanks for your suggest. I change `debug_with_location` 's function define.
```diff
_logging_debug = logging.debug
-def debug_with_location(message: str):
+def debug_with_location(message, *args, **kwargs):
    import inspect
    # get logging level
    logging_level = logging.getLogger().getEffectiveLevel()
    if logging_level != logging.DEBUG:
        return
    callerframerecord = inspect.stack()[1]
    frame = callerframerecord[0]
    info = inspect.getframeinfo(frame)
    # get caller function name
    caller = frame.f_code.co_name
-   _logging_debug(f"{info.filename}:{info.lineno}:{caller} {message}")
+   _logging_debug(f"{info.filename}:{info.lineno}:{caller} {message}", *args, **kwargs)
logging.debug = debug_with_location
```

**Comment by tanghaibao** [comment]
Thank you for the changes.
This doesn't seem to solve the caller info issue.

Bottom is the old code, and mentions that the debug is from the `need_update` location in `base.py`.
Top is with the fix, it seems to be still pointing to the same code location also.

image

Also, given that `logging.debug()` is widely used in this codebase, having lots of traceback information is going to make it a little crowded. I wonder if there is a better approach?

Thank you for your input - really appreciated.

**Comment by dwpeng** [comment]
At first I was trying to add the location of the call to the debug function by replacing the original debug function. The location here is only traced to the caller who called debug, i.e. `caller -> use_debug_function -> debug`. but as you said, there are too many places where the debug function is used, which makes it difficult to locate the debug caller.

I have an idea. For example the following code.
```py
def logging_debug(message, *args, **kwargs):
    """ logging's debug function """

def use_debug_function():
    logging_debug('debug message')

def target_caller():
    use_debug_message()
```

I can implement a decorator function to decorate `target_caller`, replace the logging_debug function only in the caller function by my debug function, in the decorator function, I always trace `my_debug` function's previous caller.

```py
def logging_debug(message, *args, **kwargs):
    """ logging's debug function """

def use_debug_function():
    logging_debug('debug message')

def my_debug(message, *args, **kwargs):
    """ trace caller and print debug message """

def patch_debug(func):
    old_debug = logging.debug
    logging.debug = my_debug
    def warp(*args, **kwargs):
        ret = func(*args, **kwargs)
        return ret
    return wrap

@patch_debug
def target_caller():
    use_debug_message()
``` 
I have written a small toy with a similar function before, you can take a look.
https://github.com/dwpeng/xprint

**Comment by dwpeng** [comment]
I write a demo.
```py
import logging
import functools
import threading

__logging_locker = threading.Lock()


def patch_debug(func):
    @functools.wraps(func)
    def wraps(*args, **kwargs):
        import inspect
        callerframerecord = inspect.stack()[1]
        frame = callerframerecord[0]
        info = inspect.getframeinfo(frame)
        # get caller function name
        caller = frame.f_code.co_name
        patch_message = f"{info.filename}:{info.lineno}:{caller}"
        old_debug = logging.debug

        def my_debug(message: str, *args, **kwargs):
            old_debug(f'{patch_message} {message}', *args, **kwargs)

        __logging_locker.acquire()
        logging.debug = my_debug
        ret = func(*args, **kwargs)
        logging.debug = old_debug
        __logging_locker.release()

        return ret

    return wraps


@patch_debug
def use_debug_function(path: str):
    logging.debug(path)


def caller():
    use_debug_function('/home')  # Line 40


if __name__ == '__main__':
    logging.basicConfig(
        format='[%(levelname)s]%(message)s',
        level=logging.DEBUG,
    )
    caller()
    # [DEBUG]/home/deng/python/class/class.py:40:caller /home
```

**Comment by dwpeng** [comment]
I write a demo.
```py
import logging
import functools
import threading

__logging_locker = threading.Lock()


def patch_debug(func):
    @functools.wraps(func)
    def wraps(*args, **kwargs):
        import inspect
        callerframerecord = inspect.stack()[1]
        frame = callerframerecord[0]
        info = inspect.getframeinfo(frame)
        # get caller function name
        caller = frame.f_code.co_name
        patch_message = f"{info.filename}:{info.lineno}:{caller}"
        old_debug = logging.debug

        def my_debug(message: str, *args, **kwargs):
            old_debug(f'{patch_message} {message}', *args, **kwargs)

        __logging_locker.acquire()
        logging.debug = my_debug
        ret = func(*args, **kwargs)
        logging.debug = old_debug
        __logging_locker.release()

        return ret

    return wraps


@patch_debug
def use_debug_function(path: str):
    logging.debug(path)


def caller():
    use_debug_function('/home')  # Line 40


if __name__ == '__main__':
    logging.basicConfig(
        format='[%(levelname)s]%(message)s',
        level=logging.DEBUG,
    )
    caller()
    # [DEBUG]/home/deng/python/class/class.py:40:caller /home
```

**Review by tanghaibao — APPROVED** [review]

**Comment by tanghaibao** [comment]
Awesome. thank you for the contribution!

**Pull request merged**